### PR TITLE
feat: warn when apply_token_bitmask_inplace leaves logits unmasked (undersized bitmask)

### PR DESCRIPTION
**Pull request by whycoming**
## Summary

Addresses the crash class reported in #611 (and upstream mlc-ai/web-llm#807).

`apply_token_bitmask_inplace(logits, bitmask)` auto-detects the vocab size as
`min(logits.shape[-1], bitmask.shape[-1] * 32)` when `vocab_size` is not passed. If the bitmask
is narrower than the logits, the trailing logits `[bitmask.shape[-1] * 32, logits.shape[-1])` are
**silently left unmasked**, so out-of-grammar tokens there (e.g. padded/reserved vocabulary ids)
can still be sampled and are then rejected/aborted by the matcher.

This is exactly what happens in #611: a caller allocates the bitmask with the tokenizer length
(Qwen3 = 151669) while the model emits logits of the padded width (151936). Token id 151757 sits
in the unmasked tail, gets sampled, and triggers `Invalid token id 151757 for GrammarMatcher` /
`Grammar matcher rejected the newly sampled token`.

Minimal repro:

```python
import torch, xgrammar as xgr

logits = torch.ones(1, 151936)                   # model's padded logit width
bitmask = xgr.allocate_token_bitmask(1, 151669)  # sized from tokenizer length (the bug)
bitmask.fill_(0)
xgr.apply_token_bitmask_inplace(logits, bitmask)
assert torch.isfinite(logits[0, 151757])         # reserved id left unmasked -> sampleable
```

## Change

- Emit a `UserWarning` from `apply_token_bitmask_inplace` when `vocab_size is None` **and**
  `bitmask.shape[-1] * 32 < logits.shape[-1]`, pointing the caller to size the bitmask to the
  model's full (padded) vocabulary (or pass the real, unpadded `vocab_size` to intentionally mask
  only a prefix).
- Document the pitfall in the "Padding" note of the docstring.

**No change to masking behavior.** The "logits wider than bitmask" path is intentionally supported
and left untouched; the warning is purely diagnostic and is skipped when `vocab_size` is passed
explicitly. It fires at most once per call site per process (Python's default warning dedup), and
keys only on the vocab dimension, so dynamic batch sizes do not defeat the dedup.

## Tests

Added to `tests/python/test_token_bitmask_operations.py`:

- `test_apply_token_bitmask_inplace_warns_on_undersized_bitmask` and `..._batched` — warns, and
  masking is unchanged (the tail stays finite).
- `test_apply_token_bitmask_inplace_no_warn_when_bitmask_covers_logits` (parametrized over wider /
  exactly-covering) — no warning at or above exact coverage.
- `test_apply_token_bitmask_inplace_no_warn_when_vocab_size_given` — explicit `vocab_size`
  suppresses the warning.

All pass on the CPU backend.

## Open question for maintainers

Since the wider-logits path is a supported feature, I kept this as a **diagnostic warning only**.
If you'd prefer it be doc-only, or gated behind an opt-in strict flag / opt-out kwarg, I'm happy to
adjust.

### CODE REVIEW

File: python/xgrammar/matcher.py [hunk 1]
```diff
@@ -146,6 +150,20 @@ def apply_token_bitmask_inplace(
             + f"But got logits.device: {logits.device}, bitmask.device: {bitmask.device}"
         )
 
+    if vocab_size is None:
+        masked_width = bitmask.shape[-1] * 32
+        if masked_width < logits.shape[-1]:
+            warnings.warn(
+                f"The bitmask covers fewer tokens (bitmask.shape[-1] * 32 = {masked_width}) than "
```
Comment (Seven-Streams) [inline python/xgrammar/matcher.py]: Maybe the warning could be shorter

**Comment by Seven-Streams** [comment]
Thanks for your contribution! I think the PR can be merged after the comment is resolved.

**Review by Seven-Streams — APPROVED** [review]
